### PR TITLE
SCHEDULES: As Ting, I would like to be able to schedule pipelines to run automatically at a suitable time and frequency, so I can safely set and forget loads of regular harvests

### PR DESCRIPTION
**Pull request by richardmatthewsdev**
**Acceptance Criteria**
- Pipelines can be scheduled
- Pipelines and definitions that are scheduled have clear visual indications and warnings (to prevent mistakenly messing with live/important definitions)

**Notes**
- Remember the last run settings when creating a new schedule/job. 
- Harvesters can make an informed decision about timing (to avoid overloading any timeslot)
- Consider transparency of timezone when scheduling

### CODE REVIEW

File: app/views/schedules/edit.html.erb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    </button>
+
+    <%= link_to 'Cancel', pipeline_schedule_path(@pipeline, @schedule), class: 'btn btn-outline-danger' %>
+  </div>
+
+  <div class="clearfix"></div>
+
+<% end %>
+
+<div class='clearfix'></div>
```
Comment (paul-mesnilgrente) [inline app/views/schedules/edit.html.erb:20]: Do you need a clearfix at the beginning of the template?

File: app/controllers/schedules_controller.rb [hunk 2]
```diff
@@ -0,0 +1,72 @@
+# frozen_string_literal: true
+
+class SchedulesController < ApplicationController
+  include LastEditedBy
```
[the remark this answers is not in the document]
Comment (richardmatthewsdev) [reply app/controllers/schedules_controller.rb:4]: Yeah it's not on the design, I'll remove it for now and we can add it if it's wanted :)

File: app/frontend/js/schedules.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  scheduleFrequencySelect.addEventListener("change", (event) => {
+    scheduleFrequencyLogic(event.target.value);
+  });
+}
+
+function scheduleFrequencyLogic(frequency) {
+  hideDynamicFrequencyInputs();
+
+  if (frequency == "daily") {
+    scheduleRunLabel.innerHTML = "Run daily at";
```
[the remark this answers is not in the document]
Comment (richardmatthewsdev) [reply app/frontend/js/schedules.js:27]: fancy 🍡

File: app/frontend/js/schedules.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  } else if (frequency == "weekly") {
+    scheduleRunLabel.innerHTML = "Run weekly at";
+    scheduleWeekly.style.display = "inline-block";
```
[the remark this answers is not in the document]
Comment (richardmatthewsdev) [reply app/frontend/js/schedules.js:30]: Yeah that works too! Good idea 🍡

File: app/frontend/stylesheets/blocks/_schedule.scss [hunk 5]
```diff
@@ -0,0 +1,3 @@
+#js-schedule-weekly, #js-schedule-monthly, #js-schedule-bi-monthly {
+  display: none;
+}
```
[the remark this answers is not in the document]
Comment (richardmatthewsdev) [reply app/frontend/stylesheets/blocks/_schedule.scss:3]: All done 🍡 .

File: app/models/schedule.rb [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+  with_options presence: true, if: :bi_monthly? do
+    validates :bi_monthly_day_one
+    validates :bi_monthly_day_two
+  end
+
+  validates :day_of_the_month, presence: true, if: -> { monthly? }
+
+  after_create  :create_sidekiq_cron_job
+  after_update  :refresh_sidekiq_cron_job
+  after_destroy :delete_sidekiq_cron_job
```
[the remark this answers is not in the document]
Comment (richardmatthewsdev) [reply app/models/schedule.rb:29]: Yeah I was in two minds about this as well, I put it here so whenever a schedule is made it makes the cron as it's kind of useless by itself. Would you prefer it to be in the controller?

File: app/views/schedules/edit.html.erb (hunk 1 above)
Comment (richardmatthewsdev) [reply app/views/schedules/edit.html.erb:20]: No we don't need both! It must have been a copy paste haha.

File: app/models/pipeline_job.rb [hunk 7]
```diff
@@ -8,6 +8,8 @@ class PipelineJob < ApplicationRecord
   belongs_to :pipeline
   belongs_to :extraction_job, optional: true
   belongs_to :destination
+  belongs_to :schedule, optional: true
+  belongs_to :user, optional: true
```
[the remark this answers is not in the document]
Comment (richardmatthewsdev) [reply app/models/pipeline_job.rb:12]: I think user is good as it reflects the model name :)

[the remark this answers is not in the document]
Comment (paul-mesnilgrente) [reply app/models/pipeline_job.rb:12]: Should we rename all the "last_edited_by" fields to make it consistent?

File: app/models/schedule.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (paul-mesnilgrente) [reply app/models/schedule.rb:29]: Yeah I think so, just for consistency

[the remark this answers is not in the document]
Comment (richardmatthewsdev) [reply app/models/schedule.rb:29]: Cool I will update this :D

**Review by motizuki — APPROVED** [review]
Looks good guys!
if you could add a screenshot next time for context, it would be great!

File: app/models/pipeline_job.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (richardmatthewsdev) [reply app/models/pipeline_job.rb:12]: hmm that feels like a lot of change. I will update this user to be launched by to follow the pattern that you have done.

**Comment by richardmatthewsdev** [comment]
Thanks @motizuki ! Yeah sounds good 🍡

**Pull request merged**